Set process title when running cron job

fredden opened the pull request:
### Description
When managing a busy server, it's helpful to understand what each process is doing. Currently there is no visibility for a systems administrator to see what task the `bin/magento cron:run` process is currently performing. The most information that's available is that some job (or set of jobs) is running within a PHP process that was started via the system task scheduler (read: cron daemon). Because of the way that Magento spawns additional processes for some groups (depending on their configuration) it is possible to sometimes narrow down the list of possible tasks because the group name is listed in the command arguments.

This pull request sets the title of the currently-running process to clearly list the group and job names.

<details><summary>Before screen-shot</summary>

![Screenshot_2021-10-04_22-59-50](https://user-images.githubusercontent.com/334786/135930916-0325b3a5-609b-4831-a534-f74fca30298a.png)

</details><details><summary>After screen-shot</summary>

![Screenshot_2021-10-04_22-56-17](https://user-images.githubusercontent.com/334786/135930903-022cf1d7-87a5-48ee-9250-83bd1bb7576d.png)

</details>

### Related Pull Requests
*None*
<!-- related pull request placeholder -->

### Manual testing scenarios
1. Run the standard Magento cron with `bin/magento cron:run` when several long-running tasks are scheduled
1. Observe lack of (or presence of after this change) information about which job is running in the process tree (using tools like `ps` or `htop` on Unix - untested on Windows, but the functionality should work there too)

### Questions or comments
Should the path to Magento be listed in the process title too? If there are multiple Magento instances on a single server, they should be running under different users. I can't think of a sensible non-development environment where multiple Magento instances run under the same system user.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#34321: Set process title when running cron job

fredden committed:
Set process title when running cron job

fredden commented:
@magento run all tests

hostep commented:
@fredden: very nice, this sounds super useful!

Can you put a screenshot here, of how it looks with `ps` or `top`/`htop` ?

> Should the path to Magento be listed in the process title too? If there are multiple Magento instances on a single server, they should be running under different users. I can't think of a sensible non-development environment where multiple Magento instances run under the same system user.

We do occasionally have this situation, it's usually for some demo shops or something like that, never for a real production shop, so there are still valid cases I believe.

fredden commented:
Thanks for your input @hostep. I've added before & after pictures to the main pull request description. I'm still somewhat undecided about what the process title should contain; I'm confident that including the process group & current job name will be helpful. I don't know if there's enough room for a full path too. Perhaps the path should go at the end?

hostep commented:
Looks really useful, but it might introduce backwards incompatibilities, for example, [the following code](https://github.com/magento/ece-tools/blob/03a900824e1600a40dbfef1d4128840b91929a87/src/Util/BackgroundProcess.php#L49-L56) will probably stop working ...

If you could somehow try to append the extra info to the original output, it might be better?
For example:
```
/usr/local/bin/php /var/www/magento2/bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1 (group: index, job: indexer_reindex_all_invalid)
```

But not sure how complicated this would be, not sure if you can fetch the original output using `cli_get_process_title` to be able to append the extra info?

fredden committed:
Preserve (most of the) original command line title

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into cron-process-title

fredden committed:
Correct docblock

fredden commented:
Thanks for your feedback, Pieter. Based on this I've updated the pull request to preserve the original command line arguments and add the new information at the end. I've updated the screen-shots in the main description.

I'm open to including `PHP_BINARY` at the start of the new title; let me know if that'd be useful too.

fredden commented:
@magento run all tests

fredden committed:
Include PHP_BINARY in process title

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into cron-process-title

Den4ik commented:
@magento run all tests

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into cron-process-title

fredden commented:
@magento run all tests

fredden committed:
Fix unit test

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into cron-process-title

Den4ik commented:
@magento create issue

sidolov commented on app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php:
```diff
@@ -128,6 +129,16 @@ class ProcessCronQueueObserver implements ObserverInterface
+    /**
+     * @var Environment
+     */
+    protected Environment $environment;
```
PLease, make property private

sidolov commented on app/code/Magento/Cron/Observer/ProcessCronQueueObserver.php:
```diff
@@ -128,6 +129,16 @@ class ProcessCronQueueObserver implements ObserverInterface
+    /**
+     * @var string
+     */
+    protected string $originalProcessTitle;
```
PLease, make property private

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into cron-process-title

fredden committed:
Reduce property visibility

engcom-Charlie committed:
Merged 2.4-develop and resolved conflicts

engcom-Charlie committed:
Merge remote-tracking branch 'magento2/2.4-develop' into cron-process-title

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
@magento run Functional Tests EE, Functional Tests B2B, Functional Tests CE

fredden committed:
Merge remote-tracking branch 'upstream/2.4-develop' into cron-process-title

fredden committed:
Recover from failed merge conflict resolution

fredden committed:
Extract new feature to its own method

fredden committed:
Set required visibility on consts

fredden commented:
@magento run all tests

engcom-Charlie committed:
Merge remote-tracking branch 'magento2/2.4-develop' into cron-process-title

engcom-Charlie committed:
Merge branch 'cron-process-title' of github.com:fredden/magento2 into cron-process-title

engcom-Charlie commented:
@magento run all tests

engcom-Charlie committed:
Merge remote-tracking branch 'magento2/2.4-develop' into cron-process-title

engcom-Charlie commented:
@magento run all tests

engcom-Charlie commented:
@magento run Magento Health Index, Database Compare, Functional Tests CE, Functional Tests B2B, Functional Tests EE, Integration Tests

Den4ik commented:
Hi @fredden. Is it possible don't use global variables at `Test` file?

fredden committed:
Remove use of superglobal

fredden commented:
@magento run all tests

engcom-Charlie commented:
@magento run Functional Tests B2B,Functional Tests EE

engcom-Charlie commented:
Hi @fredden,

Thank you for your contribution!

The test cases which are failing now are not part of this PR or not failing because of the PR changes. 

It seems, the requested changes have been taken care so requesting for the review hence moving the PR in Pending Review.

Thank you!

fredden commented:
@magento run Functional Tests B2B, Functional Tests EE

andrewbess committed:
Merge branch '2.4-develop' into cron-process-title

fredden committed:
Merge branch '2.4-develop' into cron-process-title

fredden commented:
@magento run all tests

engcom-Charlie commented:
It seems that bot moved this PR from Pending Review to Changes Requested hence moving it back for review.

Thank you!

Den4ik commented:
@magento run Functional Tests EE, Functional Tests B2B, Functional Tests CE

engcom-Echo committed:
Merge branch '2.4-develop' into cron-process-title

engcom-Echo commented:
@magento run all tests

fredden commented:
@magento run Functional Tests CE

engcom-Echo commented:
As on last two Functional Tests CE  run `MC-14271: Admin user is not able to delete the own account` test is failing hence moving to `Extended Testing` to check the build failures.

engcom-Echo commented:
Functional Tests CE failure are different on last two run.
Run1
![image](https://user-images.githubusercontent.com/51681435/219392382-600dafd0-ffdd-4951-8fc2-0d66b41f2a74.png)

Run2
![image](https://user-images.githubusercontent.com/51681435/219392382-600dafd0-ffdd-4951-8fc2-0d66b41f2a74.png)

Hence moving to `Merge In Progress`.
